### PR TITLE
Introduce basic `Instance` type

### LINKED ISSUE #132: Introduce `Instance` structs
They can be validated with `Schema` :+1:

### PR DESCRIPTION
**Pull request by sandreae**
# `Instance`

>  A mutable representation of the reduced state of a collection of `Operations`.

A quite thin wrapper type around `HashMap` with conversion traits from `Operation`. Can only be instantiated from a CREATE `Operation`. UPDATE operations can be applied through an `update()` method. Can be instantiated with validation via `Schema.instance_from_create()` method. Eventually a lot of this may be abstracted away into `Document` methods.

### This PR

- [x] Derive an `Instance` from a CREATE  `Operation` using `TryFrom` conversion trait
- [x] Update an `Instance` with an UPDATE `Operation`
- [x] Create and validate an `Instance` with a method on `Schema` and passing in a CREATE `Operation`
- [x] ?? Maybe this is actually it for now :man_shrugging: ??

### For later
- `wasm` bindings
- the rest of  [tracking issue](https://github.com/p2panda/p2panda/issues/158)

### `Instance::try_from()`
```rust
// Convert a CREATE `Operation` into an `Instance`
let instance = Instance::try_from(create_operation)?;
```

### `Schema.instance_from_create()`
```rust
let chat_schema_defnition = "chat = { (
    message: { type: \"str\", value: tstr }
) }";

let chat = Schema::new(&schema_hash, &chat_schema_defnition.to_string())?;

let chat_instance = chat.instance_from_create(chat_create_operation)?;
```

### `update()`

```rust
let instance = Instance::try_from(create_operation)?;
instance.update(update_operation)?;
```
See [tracking issue](https://github.com/p2panda/p2panda/issues/158) for materialisation overview

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Do descriptions and terminology match `handbook` content (and visa-versa) 

## Dependencies
depends on #156 
closes #132

### CODE REVIEW

**Review by adzialocha — APPROVED** [review]

**Review by adzialocha — CHANGES_REQUESTED** [review]
Looking good! Only the Changelog entry is missing ..

Also, I'm not sure about the name `Instance` yet .. (`DocumentView` is nice), but until we haven't decided we keep it for now.

File: p2panda-rs/src/instance/instance.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+pub struct Instance(HashMap<String, OperationValue>);
+
+impl Instance {
+    /// Instantiate a new `Instance`
+    fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    /// Update this `Instance` from an UPDATE `Operation`
+    pub fn update(&mut self, operation: Operation) -> Result<(), InstanceError> {
```
Comment (adzialocha) [inline p2panda-rs/src/instance/instance.rs:22]: Another good name for this method could be `process_update` or `apply_update`

**Comment by sandreae** [comment]
There's a changelog entry in the added section, is that not where you expected it?

**Comment by adzialocha** [comment]
> There's a changelog entry in the added section, is that not where you expected it?

Pardon, I somehow didn't see it 😇

**Review by adzialocha — APPROVED** [review]

File: p2panda-rs/src/instance/instance.rs (hunk 1 above)
Comment (sandreae) [reply p2panda-rs/src/instance/instance.rs:22]: I like these names!!

**Pull request merged**